cal-adapt/climakitae
# Bug fix: threshold tools nonfinite value bug when calculating 1-in-X precipitation events in arid regions

# Description of PR
IOU informed us that for specific arid/desert locations when running 1-in-100 precipitation event with LOCA produces a “non-finite value” error, which breaks the code and stops the run. The issue is caused by nan values being produced in the threshold_tools.get_block_maxima function, specifically in the groupby option for selecting max/min when calculating the ams series prior to calculating return values that is passed to vulnerability.cava_data._metric_agg(). 

This PR: 
- Implemented a patch fix just for precipitation by setting the NaN fill value to 0 in the max scenario (for max precip, 0 will always be the minimum – this is a semi-safe solution, but not 100% accurate), and by setting the NaN fill value to the precipitation max value in the min scenario (for min precip, we cannot set to 0 since that will be the minimum)
- Included a few warning statements, and in the case of other missing data for other variables,  set the return value to 0 so the code would continue running, but warned the user that this may not be accurate and to notify the team
   - I did not evaluate why there was a missing value in the annual time series. My hypothesis is that for one specific GWL year the removal of trace precipitation (10e-10) produced a full year of nans that persisted. 

What is needed next:
- A more thorough evaluation of missing values in the datarray series resampling and how to handle. 
- Ideally we should be able to skip over them and calculate the nanmax/nanmin rather than setting a fill value. 
- This should be tested in all of the resampling scenarios done in this function. Fix was only implemented for groupby options min and max.
- This needs to be tested in `cava_data` and in threshold_tools as a whole. 

What was attempted and failed:
- skipna(True), min_count=1, fillna(), nanmax, nanmin
- It appears that our version of xarray (2022.x.x) is preventing us from using some of the more modern fixes to this issue – should also be addressed and updated, if possible.

### Summary of changes and related issue
**How to test**
In the `vulnerability_assessment` notebook, test the following code on these locations:

You will need to create a pandas dataframe of these locations with `lat` and `lon` columns
34.765625, -115.390625
33.265625, -114.796875
33.265625, -114.734375
33.234375, -115.109375

```
data = cava_data(
    ## Set-up
    sce_locs,
    downscaling_method="Statistical",  # LOCA2 data
    approach="Warming Level", 
    warming_level=2.0,

    ## 1-in-X event specific arguments
    variable="Precipitation (total)",
    metric_calc="max", # daily maximum precipitation
    one_in_x=100, # One-in-X
    units="inches", # change units

    ## Export
    export_method="None",
    file_format="NetCDF"
    # batch_mode=True # batch mode for warming level still in development
)
```
Ensure that the `nonfinite value` warning does not come up and that the run completes. 

### Relevant motivation and context
This bug was breaking the `cava_data` run for LOCA data for 1-in-X precipitation events, specifically in arid regions. 

### Type of change

## Definition of Done Checklist

#### Practical
- [N/A] 80% unit test coverage
- [N/A] Documentation
  - [N/A] All functions/adjusted functions documented in the [readthedocs](https://climakitae.readthedocs.io/en/latest/).
  - [N/A] Documentation is pushed
  - [N/A] Helper functions hidden with `_` before the name
- [x] Context of function is clearly provided
  - [x] Intent of function is provided
  - [x] How to test, so that it is not siloed on scientists and anyone can review
  - [x] Appropriate manual testing was completed
- [x] Linting completed and resolved

#### Conceptual
- [ ] Doesn't replicate existing functionality
- [ ] Aligns with general coding standard of existing functions
- [x] Matches desired functionality from users/scientists
